spacy/matcher/dependencymatcher.pyx

```diff
@@ -253,146 +298,137 @@ cdef class DependencyMatcher:
-        for patternNode in allPatternNodes:
-            self.recurse(tree, id_to_position, _node_operator_map, patternLength+1, visited_nodes+[patternNode], matched_trees)
-
-    # Given a node and an edge operator, to return the list of nodes
-    # from the doc that belong to node+operator. This is used to store
-    # all the results beforehand to prevent unnecessary computation while
-    # pattern matching
-    # _node_operator_map[node][operator] = [...]
-    def get_node_operator_map(self, doc, tree, id_to_position, nodes, root):
-        _node_operator_map = {}
-        all_node_indices = nodes.values()
-        all_operators = []
-        for node in all_node_indices:
-            if node in tree:
-                for child in tree[node]:
-                    all_operators.append(child[INDEX_RELOP])
-        all_operators = list(set(all_operators))
-        all_nodes = []
-        for node in all_node_indices:
-            all_nodes = all_nodes + id_to_position[node]
-        all_nodes = list(set(all_nodes))
-        for node in all_nodes:
-            _node_operator_map[node] = {}
-            for operator in all_operators:
-                _node_operator_map[node][operator] = []
-        for operator in all_operators:
-            for node in all_nodes:
-                _node_operator_map[node][operator] = self._ops.get(operator)(doc, node)
-        return _node_operator_map
-
-    def dep(self, doc, node):
+    def _get_matches(self, doc, tree, tokens_to_key, keys_to_position):
+        cdef bint is_valid
+
+        all_positions = [keys_to_position[key] for key in tokens_to_key]
+
+        # Generate all potential matches by computing the cartesian product of all
+        # position of the matched tokens
+        for candidate_match in product(*all_positions):
+
+            # A potential match is a valid match if all relationhips between the
+            # matched tokens are satisfied.
+            is_valid = True
+            for left_idx in range(len(candidate_match)):
+                is_valid = self._check_relationships(doc, candidate_match, left_idx, tree)
+
+                if not is_valid:
+                    break
+
+            if is_valid:
+                yield list(candidate_match)
+
+    def _check_relationships(self, doc, candidate_match, left_idx, tree):
+        # Position of the left operand within the document
+        left_pos = candidate_match[left_idx]
+
+        for relop, right_idx in tree[left_idx]:
+
+            # Position of the left operand within the document
+            right_pos = candidate_match[right_idx]
+
+            # List of valid right matches
+            valid_right_matches = self._resolve_node_operator(doc, left_pos, relop)
+
+            # If the proposed right token is not within the valid ones, fail
+            if right_pos not in valid_right_matches:
+                return False
+
+        return True
+
+
+    @lru_cache(maxsize=None)
```

Comment:
I'm a bit nervous about this. I've had a couple of bad experiences with the code in functools, I think the module doesn't really consider all of the production considerations.

I have some specific worries here:

* `maxsize=None` is okay given that we ask for the cache to be cleared, and it's an internal method...But we can easily end up with a memory leak here if we mess it up later.
* Does the lru_cache decorator mess up pickling? Method decorators often have that problem.
* Every time we call into this, it will have to call `hash(doc)` in order to make the key to the cache. I think currently this just uses `id(doc)` so it's fast, but we've debated changing it and it shouldn't be necessary.

I suggest we remove the `lru_cache` decorator and just do the memoisation ourselves:

```python

def _resolve_node_operator(self, cache, doc, node, operator): 
    key = (node, operator)
    if key not in cache:
        cache[key] = [token.i for token in self._ops[operator](doc, node)]
    return cache[key]
```

This makes the cache a local variable of the scope that's calling into this function, rather than a global, so it shouldn't be possible to leak memory even if we screw up later.

Reply:
Updated